File: samples/TeamsSDK/Archived/graph-proactive-installation/python/bots/proactiveBot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import asyncio

from botbuilder.core import TurnContext, MessageFactory
from botbuilder.core.teams import TeamsInfo, TeamsActivityHandler
from botbuilder.schema import ConversationParameters

from config import DefaultConfig
from Models.Proactive_App_Installation_Helper import ProactiveAppInstallationHelper


class ProactiveBot(TeamsActivityHandler):
    def __init__(self, config: DefaultConfig):
        super().__init__()
        self.config = config
        self.conversation_references = {}

    async def on_conversation_update_activity(self, turn_context: TurnContext):
        self._add_conversation_reference(turn_context.activity)

    async def on_teams_members_added(self, members_added, turn_context: TurnContext):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                self._add_conversation_reference(turn_context.activity)

    async def on_message_activity(self, turn_context: TurnContext):
        TurnContext.remove_recipient_mention(turn_context.activity)
        text = turn_context.activity.text.strip().lower()

        if "install" in text:
            await self._install_app_in_personal_scope(turn_context)
        elif "send" in text:
            await self._send_notification_to_all_users(turn_context)

    async def _install_app_in_personal_scope(self, turn_context: TurnContext):
        new_app_install_count = 0
        existing_app_install_count = 0

        helper = ProactiveAppInstallationHelper(self.config)
        team_members = await TeamsInfo.get_paged_members(turn_context)

        tasks = [
            helper.install_app_in_personal_scope(
                turn_context.activity.conversation.tenant_id, member.aad_object_id
            )
            for member in team_members.members
            if member.aad_object_id not in self.conversation_references
        ]

        results = await asyncio.gather(*tasks)

        for status_code in results:
            if status_code == 409:
                existing_app_install_count += 1
            elif status_code == 201:
                new_app_install_count += 1

        await turn_context.send_activity(
            MessageFactory.text(
                f"Existing: {existing_app_install_count}\n\nNewly Installed: {new_app_install_count}"
            )
        )

    async def _send_notification_to_all_users(self, turn_context: TurnContext):
        team_members = await TeamsInfo.get_paged_members(turn_context)
        sent_msg_count = len(team_members.members)

        service_url = turn_context.activity.service_url
        channel_id = turn_context.activity.channel_id
        bot_app_id = self.config.APP_ID
        tenant_id = turn_context.activity.conversation.tenant_id

        for member in team_members.members:
            conversation_parameters = ConversationParameters(
                is_group=False,
                bot=turn_context.activity.recipient,
                members=[member],
                tenant_id=tenant_id,
            )

            async def _callback(context: TurnContext):
                await context.send_activity("Proactive hello.")

            await turn_context.adapter.create_conversation(
                bot_app_id,
                _callback,
                conversation_parameters,
                channel_id,
                service_url,
            )

        await turn_context.send_activity(
            MessageFactory.text(f"Message sent: {sent_msg_count}")
        )

    def _add_conversation_reference(self, activity):
        conversation_reference = TurnContext.get_conversation_reference(activity)
        self.conversation_references[
            conversation_reference.user.aad_object_id
        ] = conversation_reference

File: samples/TeamsSDK/Archived/graph-proactive-installation/nodejs/bots/proactiveBot.js
const { ActivityHandler, TurnContext, TeamsInfo, MessageFactory } = require('botbuilder');
const ProactiveAppInstallationHelper = require('../Models/ProactiveAppInstallationHelper');

class ProactiveBot extends ActivityHandler {
    constructor(conversationReferences) {
        super();
        this.conversationReferences = conversationReferences;

        this.onConversationUpdate(async (context, next) => {
            this.addConversationReference(context.activity);
            await next();
        });

        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (const member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    this.addConversationReference(context.activity);
                }
            }
            await next();
        });

        this.onMessage(async (context, next) => {
            TurnContext.removeRecipientMention(context.activity);
            const text = (context.activity.text || '').trim().toLocaleLowerCase();
            if (text.includes('install')) {
                await this.installAppInTeamsAndChatMembersPersonalScope(context);
            } else if (text.includes('send')) {
                await this.sendNotificationToAllUsersAsync(context);
            }
            await next();
        });
    }

    async installAppInTeamsAndChatMembersPersonalScope(context) {
        let newAppInstallCount = 0;
        let existingAppInstallCount = 0;
        const helper = new ProactiveAppInstallationHelper();
        const teamMembers = await TeamsInfo.getPagedMembers(context);

        const results = await Promise.all(
            teamMembers.members.map(async (member) => {
                if (!this.conversationReferences[member.aadObjectId]) {
                    return helper.installAppInPersonalScope(context.activity.conversation.tenantId, member.aadObjectId);
                }
                return null;
            })
        );

        for (const statusCode of results) {
            if (statusCode === 409) existingAppInstallCount++;
            else if (statusCode === 201) newAppInstallCount++;
        }

        await context.sendActivity(MessageFactory.text(`Existing: ${existingAppInstallCount} \n\n Newly Installed: ${newAppInstallCount}`));
    }

    async sendNotificationToAllUsersAsync(context) {
        const teamMembers = await TeamsInfo.getPagedMembers(context);
        const sentMsgCount = teamMembers.members.length;

        for (const member of teamMembers.members) {
            const conversationParameters = {
                isGroup: false,
                channelData: {
                    tenant: { id: context.activity.conversation.tenantId }
                },
                bot: context.activity.recipient,
                members: [member]
            };

            await context.adapter.createConversationAsync(
                process.env.MicrosoftAppId,
                context.activity.channelId,
                context.activity.serviceUrl,
                null,
                conversationParameters,
                async (turnContext) => {
                    await turnContext.sendActivity('Proactive hello.');
                }
            );
        }

        await context.sendActivity(MessageFactory.text(`Message sent: ${sentMsgCount}`));
    }

    addConversationReference(activity) {
        const conversationReference = TurnContext.getConversationReference(activity);
        this.conversationReferences[conversationReference.user.aadObjectId] = conversationReference;
    }
}

module.exports.ProactiveBot = ProactiveBot;